From bd8123a7cb266fc2a5d61353efcb6a9bde8bd303 Mon Sep 17 00:00:00 2001
From: UniFrog CI <ci@unifrog.local>
Date: Sun, 10 May 2026 23:38:47 +0000
Subject: [PATCH] Add SF2000 gpsp_multicore safety hooks

Enable in-memory ROM loading for the SF2000 host, add load progress callbacks, and carry the gpSP MIPS dynarec safety guards into gpsp_multicore.
---
 common.h            |  9 +++++++
 cpu_threaded.c      | 12 ++++++---
 gba_memory.c        | 66 +++++++++++++++++++++++++++++++++++++++++++--
 libretro/libretro.c | 11 ++++++++
 main.c              |  7 ++++-
 mips/mips_emit.h    | 18 ++++++++++---
 mips/mips_stub.S    | 47 ++++++++++++++++++++------------
 7 files changed, 142 insertions(+), 28 deletions(-)

diff --git a/common.h b/common.h
index 690e5b6..a02833c 100644
--- a/common.h
+++ b/common.h
@@ -78,6 +78,15 @@
   typedef signed long long int s64;
 #endif
 
+#ifdef SF2000
+#define SF2000_GPSP_DYNAREC_LOOKUP_FAILED 0x7ffffffeu
+void unifrog_core_load_progress(const char *stage, unsigned current, unsigned total);
+#define sf2000_load_progress(stage, current, total) \
+  unifrog_core_load_progress((stage), (unsigned)(current), (unsigned)(total))
+#else
+#define sf2000_load_progress(stage, current, total) ((void)0)
+#endif
+
 #if defined(USE_XBGR1555_FORMAT)
   #define convert_palette(value)  \
     (value & 0x7FFF)
diff --git a/cpu_threaded.c b/cpu_threaded.c
index 13331ae..94b193a 100644
--- a/cpu_threaded.c
+++ b/cpu_threaded.c
@@ -2657,13 +2657,15 @@ u8 function_cc *block_lookup_address_arm(u32 pc)
   unsigned i;
   for (i = 0; i < 2; i++) {
     u8 *ret = block_lookup_translate_arm(pc);
-    if (ret) {
+    if (ret && ret != (u8 *)(~0)) {
       translate_icache_sync();
       return ret;
     }
+    if (ret == (u8 *)(~0))
+      break;
   }
 
-  // PERFORMANCE: Disable logging in hot path
+  reg[REG_PC] = pc;
   return NULL;
 }
 
@@ -2672,12 +2674,14 @@ u8 function_cc *block_lookup_address_thumb(u32 pc)
   unsigned i;
   for (i = 0; i < 2; i++) {
     u8 *ret = block_lookup_translate_thumb(pc);
-    if (ret) {
+    if (ret && ret != (u8 *)(~0)) {
       translate_icache_sync();
       return ret;
     }
+    if (ret == (u8 *)(~0))
+      break;
   }
-  // PERFORMANCE: Disable logging in hot path
+  reg[REG_PC] = pc;
   return NULL;
 }
 
diff --git a/gba_memory.c b/gba_memory.c
index b34c794..4d370fd 100644
--- a/gba_memory.c
+++ b/gba_memory.c
@@ -2880,12 +2880,14 @@ static s32 load_gamepak_raw(const char *name)
 
     // Unmap the ROM space since we will re-map it now
     map_null(read, 0x8000000, 0xD000000);
+    sf2000_load_progress("ROM", 0, ldblks);
 
     // Proceed to read the whole ROM or as much as possible.
     for (i = 0; i < ldblks; i++)
     {
       // Load 1MB chunk and map it
       filestream_read(gamepak_file_large, gamepak_buffers[i], gamepak_buffer_blocksize);
+      sf2000_load_progress("ROM", i + 1, ldblks);
       for (j = 0; j < 32 && i*32 + j < rom_blocks; j++)
       {
         u32 phyn = i*32 + j;
@@ -2903,13 +2905,73 @@ static s32 load_gamepak_raw(const char *name)
   return -1;
 }
 
+static s32 load_gamepak_memory(const struct retro_game_info *info)
+{
+  unsigned i, j;
+  const u8 *rom = (const u8 *)info->data;
+  u32 source_size = (u32)info->size;
+  u32 buf_blocks;
+  u32 rom_blocks;
+
+  if (!rom || source_size == 0)
+    return -1;
+
+  gamepak_size = (source_size + 0x7FFF) & ~0x7FFF;
+  buf_blocks = (gamepak_size + gamepak_buffer_blocksize - 1) /
+    gamepak_buffer_blocksize;
+  rom_blocks = gamepak_size >> 15;
+
+  if (buf_blocks > gamepak_buffer_count)
+    return -1;
+
+  if (gamepak_file_large)
+  {
+    filestream_close(gamepak_file_large);
+    gamepak_file_large = NULL;
+  }
+
+  map_null(read, 0x8000000, 0xD000000);
+  sf2000_load_progress("ROM_MEM", 0, buf_blocks);
+
+  for (i = 0; i < buf_blocks; i++)
+  {
+    size_t src_offset = (size_t)i * gamepak_buffer_blocksize;
+    size_t copy_size = source_size > src_offset ?
+      source_size - src_offset : 0;
+
+    if (copy_size > gamepak_buffer_blocksize)
+      copy_size = gamepak_buffer_blocksize;
+    memset(gamepak_buffers[i], 0xff, gamepak_buffer_blocksize);
+    if (copy_size)
+      memcpy(gamepak_buffers[i], rom + src_offset, copy_size);
+    sf2000_load_progress("ROM_MEM", i + 1, buf_blocks);
+
+    for (j = 0; j < 32 && i * 32 + j < rom_blocks; j++)
+    {
+      u32 phyn = i * 32 + j;
+      u8 *blkptr = &gamepak_buffers[i][32 * 1024 * j];
+      u32 entry = evict_gamepak_page();
+      gamepak_blk_queue[entry].phy_rom = phyn;
+      map_rom_entry(read, phyn, blkptr, rom_blocks);
+    }
+  }
+
+  return 0;
+}
+
 u32 load_gamepak(const struct retro_game_info* info, const char *name)
 {
    char *p;
    char gamepak_filename[512];
    gamepak_info_t gpinfo;
 
-   if (load_gamepak_raw(name))
+   sf2000_load_progress("GPSP_GAMEPAK", 1, 3);
+   if (info && info->data && info->size)
+   {
+      if (load_gamepak_memory(info))
+         return -1;
+   }
+   else if (load_gamepak_raw(name))
       return -1;
 
    strncpy(gamepak_filename, name, sizeof(gamepak_filename));
@@ -2942,6 +3004,7 @@ u32 load_gamepak(const struct retro_game_info* info, const char *name)
 
    load_game_config_over(&gpinfo);
 
+   sf2000_load_progress("GPSP_GAMEPAK_DONE", 3, 3);
    return 0;
 }
 
@@ -2958,4 +3021,3 @@ s32 load_bios(char *name)
   return 0;
 }
 
-
diff --git a/libretro/libretro.c b/libretro/libretro.c
index a9001b3..a6364a4 100644
--- a/libretro/libretro.c
+++ b/libretro/libretro.c
@@ -462,7 +462,11 @@ void retro_get_system_info(struct retro_system_info* info)
  #else
    info->library_version = GPSP_VERSION;
  #endif
+   #ifdef SF2000
+   info->need_fullpath = false;
+#else
    info->need_fullpath = true;
+#endif
    info->block_extract = false;
    info->valid_extensions = "gba|bin|agb|gbz|u1" ;
 }
@@ -1006,6 +1010,7 @@ bool retro_load_game(const struct retro_game_info* info)
    if (!info)
       return false;
 
+   sf2000_load_progress("GPSP_LOAD_START", 1, 7);
    use_libretro_save_method = 0;
    check_variables(1);
    set_input_descriptors();
@@ -1030,6 +1035,7 @@ bool retro_load_game(const struct retro_game_info* info)
    else
       strcpy(filename_bios, main_path);
 
+   sf2000_load_progress("GPSP_LOAD_BIOS", 2, 7);
    bool bios_loaded = false;
    if (selected_bios == auto_detect || selected_bios == official_bios)
    {
@@ -1057,16 +1063,21 @@ bool retro_load_game(const struct retro_game_info* info)
    }
 
    memset(gamepak_backup, -1, sizeof(gamepak_backup));
+   sf2000_load_progress("GPSP_LOAD_GAMEPAK", 3, 7);
    if (load_gamepak(info, info->path) != 0)
    {
       error_msg("Could not load the game file.");
       return false;
    }
+   sf2000_load_progress("GPSP_LOAD_GAMEPAK_OK", 4, 7);
 
+   sf2000_load_progress("GPSP_LOAD_RESET", 5, 7);
    reset_gba();
 
+   sf2000_load_progress("GPSP_LOAD_MEMDESC", 6, 7);
    set_memory_descriptors();
 
+   sf2000_load_progress("GPSP_LOAD_OK", 7, 7);
    return true;
 }
 
diff --git a/main.c b/main.c
index d2cc7d3..0056afb 100644
--- a/main.c
+++ b/main.c
@@ -631,10 +631,15 @@ u32 function_cc update_gba(int remaining_cycles)
 
 void reset_gba(void)
 {
+  sf2000_load_progress("GPSP_RESET_MEMORY", 1, 5);
   init_memory();
+  sf2000_load_progress("GPSP_RESET_MAIN", 2, 5);
   init_main();
+  sf2000_load_progress("GPSP_RESET_CPU", 3, 5);
   init_cpu();
+  sf2000_load_progress("GPSP_RESET_SOUND", 4, 5);
   reset_sound();
+  sf2000_load_progress("GPSP_RESET_DONE", 5, 5);
 }
 
 #ifdef TRACE_REGISTERS
@@ -750,4 +755,4 @@ unsigned main_write_savestate(u8* dst)
   bson_finish_document(dst, wbptr);
 
   return (unsigned int)(dst - startp);
-}
\ No newline at end of file
+}
diff --git a/mips/mips_emit.h b/mips/mips_emit.h
index 2a19d0f..39faaff 100644
--- a/mips/mips_emit.h
+++ b/mips/mips_emit.h
@@ -2655,13 +2655,16 @@ static void emit_phand(
   mips_emit_addu(reg_temp, reg_temp, reg_base); // Add to the base_reg the table offset
   mips_emit_lw(reg_rv,   reg_temp, tbloff);     // Get func addr from 1st table
   mips_emit_lw(reg_temp, reg_temp, tbloff2);    // Get opcode from 2nd table
-  mips_emit_sw(reg_temp, mips_reg_ra, -8);      // Patch instruction!
-
   #if defined(PSP)
+    mips_emit_sw(reg_temp, mips_reg_ra, -8);    // Patch instruction!
     mips_emit_cache(0x1A, mips_reg_ra, -8);
     mips_emit_jr(reg_rv);                       // Jump directly to target for speed
     mips_emit_cache(0x08, mips_reg_ra, -8);
+  #elif defined(SF2000)
+    mips_emit_jr(reg_rv);                       // Jump directly to target for speed
+    mips_emit_nop();
   #else
+    mips_emit_sw(reg_temp, mips_reg_ra, -8);    // Patch instruction!
     mips_emit_jr(reg_rv);
     #ifdef MIPS_HAS_R2_INSTS
       mips_emit_synci(mips_reg_ra, -8);
@@ -2805,9 +2808,16 @@ void init_emitter(bool must_swap) {
 
 u32 execute_arm_translate_internal(u32 cycles, void *regptr);
 u32 execute_arm_translate(u32 cycles) {
-  return execute_arm_translate_internal(cycles, &reg[0]);
+  u32 ret = execute_arm_translate_internal(cycles, &reg[0]);
+#ifdef SF2000
+  if (ret == SF2000_GPSP_DYNAREC_LOOKUP_FAILED) {
+    clear_gamepak_stickybits();
+    execute_arm(cycles);
+    return 0;
+  }
+#endif
+  return ret;
 }
 
 #endif
 
-
diff --git a/mips/mips_stub.S b/mips/mips_stub.S
index 8310205..d04ae5e 100644
--- a/mips/mips_stub.S
+++ b/mips/mips_stub.S
@@ -18,17 +18,24 @@
 
 #include "../gpsp_config.h"
 
+#define SF2000_GPSP_DYNAREC_LOOKUP_FAILED 0x7ffffffe
+
 // This is also defined in sys/asm.h but doesn't seem portable?
 #ifdef __mips64
   .set mips64
   #define SZREG 8
   #define REG_L ld
   #define REG_S sd
-#else
+#elif defined(MIPS_HAS_R2_INSTS)
   .set mips32r2
   #define SZREG 4
   #define REG_L lw
   #define REG_S sw
+#else
+  .set mips32
+  #define SZREG 4
+  #define REG_L lw
+  #define REG_S sw
 #endif
 
 #define defsymbl(symbol) \
@@ -227,6 +234,14 @@ symbol:
 .endm
 #endif
 
+.macro jump_translated_result
+  beqz $2, translated_lookup_failed
+  addiu $1, $0, -4
+  and $2, $2, $1
+  jr $2
+  nop
+.endm
+
 
 # Process a hardware event. Since an interrupt might be
 # raised we have to check if the PC has changed.
@@ -298,22 +313,26 @@ defsymbl(mips_indirect_branch_arm)
   save_registers
   cfncall block_lookup_address_arm, 1
   restore_registers
-  jr $2                           # $2 = value returned
-  nop
+  jump_translated_result          # $2 = value returned
 
 defsymbl(mips_indirect_branch_thumb)
   save_registers
   cfncall block_lookup_address_thumb, 2
   restore_registers
-  jr $2                           # $2 = value returned
-  nop
+  jump_translated_result          # $2 = value returned
 
 defsymbl(mips_indirect_branch_dual)
   save_registers
   cfncall block_lookup_address_dual, 3
   nop
   restore_registers
-  jr $2                           # $2 = value returned
+  jump_translated_result          # $2 = value returned
+
+translated_lookup_failed:
+  collapse_flags
+  save_registers
+  li $2, SF2000_GPSP_DYNAREC_LOOKUP_FAILED
+  b return_to_main
   nop
 
 
@@ -359,13 +378,11 @@ lookup_pc_noflags:
 
   cfncall block_lookup_address_thumb, 2  # get Thumb address
   restore_registers
-  jr $2                                  # jump to result
-  nop
+  jump_translated_result                 # jump to result
 1:
   cfncall block_lookup_address_arm, 1  # get ARM address
   restore_registers
-  jr $2                                # jump to result
-  nop
+  jump_translated_result               # jump to result
 
 
 defsymbl(smc_write)
@@ -485,8 +502,7 @@ defsymbl(execute_store_cpsr)
   addu $4, $2, $0                      # load new address in $4
   cfncall block_lookup_address_arm, 1  # GBA address is in $4
   restore_registers                    # restore registers
-  jr $2                                # jump to the new address
-  nop
+  jump_translated_result               # jump to the new address
 
 
 # $4: new spsr
@@ -542,14 +558,12 @@ defsymbl(execute_arm_translate_internal)
 
   cfncall block_lookup_address_arm, 1
   restore_registers               # load initial register values
-  jr $2                           # jump to return
-  nop
+  jump_translated_result          # jump to return
 
 1:
   cfncall block_lookup_address_thumb, 2
   restore_registers               # load initial register values
-  jr $2                           # jump to return
-  nop
+  jump_translated_result          # jump to return
 
 .bss
 .align 6
@@ -629,4 +643,3 @@ defsymbl(ram_translation_cache)
   .space RAM_TRANSLATION_CACHE_SIZE
 
 #endif
-
-- 
2.54.0

